src/G2/Equiv/Induction.hs

```diff
@@ -0,0 +1,336 @@
+              (StateET, StateET) ->
+              W.WriterT [Marker] IO (Bool, StateET, IndMarker)
+inductionL solver ns prev (s1, s2) = do
+  let scr1 = innerScrutinees $ exprExtract s1
+      scr2 = innerScrutinees $ exprExtract s2
+      scr_pairs = [(sc1, sc2) | sc1 <- scr1, sc2 <- scr2]
+      scr_states = [(s1 { curr_expr = CurrExpr Evaluate sc1 }, s2 { curr_expr = CurrExpr Evaluate sc2 }) | (sc1, sc2) <- scr_pairs]
+  mr_pairs <- mapM (moreRestrictiveIndRight solver ns prev) scr_states
+  let mr_zipped = zip scr_pairs mr_pairs
+      working_info = [(sc1, sc2, pm) | ((sc1, sc2), Just pm) <- mr_zipped]
+      working_info' = filter (\(_, _, PrevMatch _ (_, p2) _ pc2) -> validScrutinee s2 p2 pc2) working_info
+  case working_info' of
+    [] -> return (False, s1, undefined)
+    h:_ -> let (sc1, sc2, PrevMatch (q1, q2) (p1, p2) (mapping, _) pc2) = h
+               e2_old = exprExtract pc2
+               hm_list = HM.toList mapping
+               e2_old' = foldr (\(i, e) acc -> replaceASTs (Var i) e acc) e2_old hm_list
+               e1_new = replaceScrutinee sc1 e2_old' $ exprExtract s1
+               h1_new = E.union (expr_env s1) (expr_env pc2)
+               s1' = s1 {
+                 curr_expr = CurrExpr Evaluate e1_new
+               , expr_env = h1_new
+               }
+               -- TODO past
+               -- TODO should result include generalization?
+               -- TODO p1, p2, q1, q2, present scrutinees
+               -- TODO do I really need q1 and q2?
+               -- TODO placeholders for real present and name to avoid undefined errors
+               im = IndMarker {
+                 ind_real_present = (s1, s2)
+               , ind_used_present = (s1, s2)
+               , ind_past = (p1, pc2)
+               , ind_result = (s1', s2)
+               , ind_present_scrutinees = (sc1, sc2)
+               , ind_past_scrutinees = (p1, p2)
+               , ind_side = ILeft
+               , ind_fresh_name = Name "" Nothing 0 Nothing
+               }
+           in return (True, s1', im)
+
+-- precedence goes to left-side substitution
+-- right-side substitution only happens if left-side fails
+-- TODO reverse prev for the right side
+-- TODO right-sided induction can happen here, even if it's called left
+induction :: S.Solver solver =>
+             solver ->
+             HS.HashSet Name ->
+             [(StateET, StateET)] ->
+             (StateET, StateET) ->
+             W.WriterT [Marker] IO (Bool, StateET, StateET, IndMarker)
+induction solver ns prev (s1, s2) = do
+  (bl, s1l, iml) <- inductionL solver ns prev (s1, s2)
+  if bl then return (bl, s1l, s2, iml)
+  else do
+    let prev' = map swap prev
+    (br, s2r, imr) <- inductionL solver ns prev' (s2, s1)
+    return (br, s1, s2r, reverseIndMarker imr)
+
+-- left side stays constant
+-- TODO complex conditional, but avoids needless generalization
+-- TODO returning marker in case of failure
+inductionFoldL :: S.Solver solver =>
```

Comment:
This and inductionFold could both use comments- what is inductionFoldL vs inductionFold?